### PR TITLE
Ability to configure client to send default headers #284

### PR DESCRIPTION
**Pull request by IlyasYOY**
Hello! 

I've tried to implement feature I requested in Issue #284 
TBH I just updated my old solution.

I've got some likes there so I decided to post this PR. 

Thanks you for your work!

### CODE REVIEW

**Comment by pivotal-issuemaster** [comment]
@IlyasYOY Please sign the [Contributor License Agreement](https://cla.pivotal.io/sign/spring?repositoryId=spring-cloud/spring-cloud-openfeign&pullRequestId=415)!

[Click here](https://cla.pivotal.io/sync/spring?repositoryId=spring-cloud/spring-cloud-openfeign&pullRequestId=415) to manually synchronize the status of this Pull Request.

See the [FAQ](https://cla.pivotal.io/about) for frequently asked questions.

**Comment by pivotal-issuemaster** [comment]
@IlyasYOY Thank you for signing the [Contributor License Agreement](https://cla.pivotal.io/sign/spring?repositoryId=spring-cloud/spring-cloud-openfeign&pullRequestId=415)!

**Comment by OlgaMaciaszek** [comment]
Thanks @IlyasYOY . Will review it tomorrow.

File: spring-cloud-openfeign-core/src/main/java/org/springframework/cloud/openfeign/FeignClientProperties.java [hunk 1]
```diff
@@ -113,6 +114,10 @@ public int hashCode() {
 
 		private ExceptionPropagationPolicy exceptionPropagationPolicy;
 
+		private Map<String, Collection<String>> defaultRequestHeaders;
+
+		private Map<String, Collection<String>> defaultRequestParameters;
```
Comment (OlgaMaciaszek) [inline spring-cloud-openfeign-core/src/main/java/org/springframework/cloud/openfeign/FeignClientProperties.java:119]: Let's rename to `defaultQueryParameters`.

File: spring-cloud-openfeign-core/src/test/java/org/springframework/cloud/openfeign/FeignClientUsingPropertiesTests.java [hunk 2]
```diff
@@ -87,6 +94,8 @@
 
 	private FeignClientFactoryBean barFactoryBean;
 
+	private FeignClientFactoryBean eggsClientFactoryBean;
```
Comment (OlgaMaciaszek) [inline spring-cloud-openfeign-core/src/test/java/org/springframework/cloud/openfeign/FeignClientUsingPropertiesTests.java:97]: Let's rename the factory bean, context and test to something meaningful, like `defaultHeadersAndQueryParams`.

File: spring-cloud-openfeign-core/src/test/resources/feign-properties.properties [hunk 3]
```diff
@@ -10,6 +10,8 @@ feign.client.config.default.retryer=org.springframework.cloud.openfeign.FeignCli
 feign.client.config.default.decode404=true
 feign.client.config.foo.requestInterceptors[0]=org.springframework.cloud.openfeign.FeignClientUsingPropertiesTests.FooRequestInterceptor
 feign.client.config.foo.requestInterceptors[1]=org.springframework.cloud.openfeign.FeignClientUsingPropertiesTests.BarRequestInterceptor
+feign.client.config.eggs.defaultRequestHeaders[eggsHeaders]=header
```
Comment (OlgaMaciaszek) [inline spring-cloud-openfeign-core/src/test/resources/feign-properties.properties:13]: Let's test also for collection values.

File: spring-cloud-openfeign-core/src/test/resources/feign-properties.properties [hunk 4]
```diff
@@ -10,6 +10,8 @@ feign.client.config.default.retryer=org.springframework.cloud.openfeign.FeignCli
...
+feign.client.config.eggs.defaultRequestParameters[eggsParameters]=parameter
```
Comment (OlgaMaciaszek) [inline spring-cloud-openfeign-core/src/test/resources/feign-properties.properties:14]: Let's test also for collection values.

File: spring-cloud-openfeign-core/src/test/java/org/springframework/cloud/openfeign/FeignClientUsingPropertiesTests.java [hunk 5]
```diff
@@ -252,6 +283,11 @@ public String bar() throws InterruptedException {
 			return "OK";
 		}
 
+		@RequestMapping(method = RequestMethod.GET, path = "/eggs")
```
Comment (OlgaMaciaszek) [inline spring-cloud-openfeign-core/src/test/java/org/springframework/cloud/openfeign/FeignClientUsingPropertiesTests.java:286]: Let's switch to `@GetMapping` to be more concise.

**Review by OlgaMaciaszek — COMMENTED** [review]
Thanks, @IlyasYOY. The functionality looks good, however, I've added a few comments.

Also, this feature does not break backward compatibility, so it could be added starting with the `Hoxton` release branch. Could you resubmit this PR (or create a new one instead) against `2.2.x` rather than `master`?

Please add your full name with `@author` tag to all the classes you've modified.

Please add a section in the documentation describing this feature (in `spring-cloud-openfeign.adoc`).

**Review by OlgaMaciaszek — CHANGES_REQUESTED** [review]
See the other comments.

**Comment by OlgaMaciaszek** [comment]
Fixes gh-284.

File: spring-cloud-openfeign-core/src/test/java/org/springframework/cloud/openfeign/FeignClientUsingPropertiesTests.java (hunk 5 above)
Comment (IlyasYOY) [reply spring-cloud-openfeign-core/src/test/java/org/springframework/cloud/openfeign/FeignClientUsingPropertiesTests.java]: Solved, also replaced all `RequestMapping` entries in this file.

File: spring-cloud-openfeign-core/src/test/resources/feign-properties.properties (hunk 4 above)
Comment (IlyasYOY) [reply spring-cloud-openfeign-core/src/test/resources/feign-properties.properties]: Added `pawsTest`, that checks that option

File: spring-cloud-openfeign-core/src/test/resources/feign-properties.properties (hunk 3 above)
Comment (IlyasYOY) [reply spring-cloud-openfeign-core/src/test/resources/feign-properties.properties:13]: Added pawsTest, that checks that option

File: spring-cloud-openfeign-core/src/test/java/org/springframework/cloud/openfeign/FeignClientUsingPropertiesTests.java (hunk 2 above)
Comment (IlyasYOY) [reply spring-cloud-openfeign-core/src/test/java/org/springframework/cloud/openfeign/FeignClientUsingPropertiesTests.java]: Renamed. `defaultHeadersAndQuerySingleParamsFeignClientFactoryBean`

File: spring-cloud-openfeign-core/src/main/java/org/springframework/cloud/openfeign/FeignClientProperties.java (hunk 1 above)
Comment (IlyasYOY) [reply spring-cloud-openfeign-core/src/main/java/org/springframework/cloud/openfeign/FeignClientProperties.java]: Renamed. `defaultQueryParameters`

**Comment by IlyasYOY** [comment]
I'll open a new PR with the changes, but applied on 2.2.x branch!

Thanks for review comments!

**Comment by IlyasYOY** [comment]
@OlgaMaciaszek Moved all the work on 2.2.x branch

You can check it here: https://github.com/spring-cloud/spring-cloud-openfeign/pull/415

I guess this PR may be closed.

**Pull request closed without merging**